Speed up frozen actor updates & prevent flicker

RoosterDragon opened the pull request:
Fixes  #7660.

Previously frozen actor updates were done twice per actor. The FrozenUnderFog tick would do this as well as the FrozenActorLayer doing it for every FrozenActor to see if the frozen actor was still visible under the current shroud conditions.

Calculating whether the actor was visible under the shroud is expensive, so we avoid doing it twice since there's no need. Now FrozenUnderFog will update everything as needed.

A couple of further commits introduce some related refactorings and cleanup.

This PR is purely optimization, the behaviour of frozen actors should remain 100% if you're testing this :)

On the RA shellmap, this reduces total CPU time spent doing the frozen actor updates from 6.21% to 4.13%.

Mailaender commented:
This regresses badly:

![image](https://cloud.githubusercontent.com/assets/756669/6426342/f7c5e958-bf55-11e4-9626-4b64bd3eb758.png)

The left pill box and wall segments have been scouted, fogged, then destroyed by artillery. Revealing them again makes them still show up in a glitchy manner.

RoosterDragon committed:
Avoid duplicated frozen visibility checks.

The FrozenUnderFog.Tick method will now reuse the calculation do by the frozen actor when it had to calculate its visibility, this prevents it having to re-do the fairly expensive visibility calculation.

RoosterDragon commented:
Yup that's broken alright - what an obvious thing to not occur to me to test!

Anyway this seems to be caused by the fact FrozenUnderFog stops ticking once the actor dies. So it we turn the logic the other way around - FrozenActorLayer does the work and FrozenUnderFog uses the result - that fixes the problem since the layer keeps on ticking.

This makes the diff a bit uglier, but the performance improvement is the same.

pchote commented:
This may also fix the flicker when switching between the regular and frozen actor rendering, which would be a great unexpected side effect.

RoosterDragon commented:
I'd never noticed that was an issue before - interesting.

Just tested it, and yes it seems I (totally unintentionally) fixed it - how fortuitous.

Mailaender commented on OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs:
```diff
@@ -66,55 +65,43 @@ public void Tick(Actor self)
-					visible[p] |= startsRevealed;
-					p.PlayerActor.Trait<FrozenActorLayer>().Add(frozen[p] = new FrozenActor(self, footprint, footprintRegion));
+					frozen[player] = frozenActor = new FrozenActor(self, footprint, footprintRegion, player.Shroud);
+					player.PlayerActor.Trait<FrozenActorLayer>().Add(frozenActor);
```
You could cache the trait lookups for all `self.World.Players` in the constructor to save even more resources.

RoosterDragon replied:
The trait lookup only happens on the first tick so it's not an issue.

Mailaender replied:
You are right.

Mailaender commented:
:+1: The regressions are fixed. Tested both revealing, fogging, destroying, revealing and starting a map without shroud and just fog where tech buildings are correctly revealed at start, but others not. :white_check_mark:

pchote commented:
I couldn't spot anything wrong with the code or ingame, so :+1: from me too.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/2887629d44b3f2083ac030ad26969a228dbaed57%5E...2887629d44b3f2083ac030ad26969a228dbaed57
